### PR TITLE
Ersuo/adding send status telemetry

### PR DESCRIPTION
**Pull request by Erli-ms**
> Related to #4500.

## Changelog Entry

- Added function to emit status change telemetry event for activities, by [@Erli-ms](https://github.com/Erli-ms), in PR [#4631](https://github.com/microsoft/BotFramework-WebChat/pull/4631)

## Description

Added an ActivityStatusTelemetryComposer inside of ActivitySendStatusComposer. Whenever the activity status changes, the ActivityStatusTelemetryComposer collects the map of concerned activities and triggers trackEvent to emit the event. Consumers of the events can react accordingly. This feature is requested by D365 OmniChannel team. 

## Design

Added a new provider to look at send status of every activity. Whenever send status of any activities changed, it will call `onTelemetry` event to fire its changes.

## Specific Changes

- packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx
    - creating a new composer triggered whenever the control got re-rendered.
    - collect the activities which status changed
    - invoke trackEvent to emit the event to the telemetry logger
    - `{ status, prevStatus, ClientActivityID, type, hasAttachment, key}` is emitted to the telemetry logger
- packages/api/src/hooks/Composer.tsx
    - added the ActivitySendStatusTelemetryComposer as a children component to ActivitySendStatusComposer
- tests/html/activityStatusTelemetry.sendingToSendFailed.html
    - added test to verify expected event been emitted with the activity status changed from 'sending' to 'send failed'
- tests/html/activityStatusTelemetry.sendingToSendFailedToSent.html
    - added test to verify expected event been emitted with the activity status changed from 'sending' to 'send failed' and from 'send failed' to 'sent'
-  tests/html/activityStatusTelemetry.sendingToSent.html
    - added test to verify expected event been emitted with the activity status changed from 'sending' to 'sent'
- tests/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html
    - added test to verify expected event been emitted with when attachment activity status changed from 'sending' to 'sent'
- tests/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html
    - added test to verify incoming activity does not emit any status change event

## -

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] Documents reviewed (docs, samples, live demo)
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const ActivitySendStatusTelemetryComposer = () => {
+  const [activityToSendStatusMap] = useSendStatusByActivityKey();
+  const previousActivityToSendStatusMap = usePrevious(activityToSendStatusMap);
+  const getActivityByKey = useGetActivityByKey();
+  const trackEvent = useTrackEvent();
+
+  if (activityToSendStatusMap && previousActivityToSendStatusMap) {
+    const allActivityKeys = activityToSendStatusMap.keys();
+
+    for (const key of allActivityKeys) {
```
Comment (compulim) [inline packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:14]: Try to eliminate one use constants if they are (subjectively) easy to read.

```suggestion
    for (const key of activityToSendStatusMap.keys()) {
```

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      const currentStatus = activityToSendStatusMap.get(key);
+      const previousStatus = previousActivityToSendStatusMap.get(key);
+
+      if (!!currentStatus && !!previousStatus && currentStatus !== previousStatus) {
```
Comment (compulim) [inline packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:18]: No need `!!` in `if`, because `if` knows truthy/falsy.

```suggestion
      // `status` is falsy if it is not an outgoing activity.
      // `prevStatus` is falsy if it is not an outgoing activity, or it is recently added.
      // This telemetry data point only emit changes in outgoing activities.
      if (status && prevStatus && status !== prevStatus) {
```

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        const activity = getActivityByKey(key);
+        const telemetryPayload = {
+          currentStatus: currentStatus.toString(),
+          previousStatus: previousStatus.toString(),
+          clientActivityID: activity?.channelData?.clientActivityID,
+          type: activity?.type?.toString(),
+          key,
+          hasAttachment: activity?.type === 'message' && activity?.attachments?.length > 0 ? 'true' : 'false'
+        };
```
Comment (compulim) [inline packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:27]: Notes:
- No need `toString()` because `type SendStatus extends string`, we don't use `enum`
   - Same to `activity.type.toString()`
- Sorting the object by alphabetical order of the key
- Saving `activity.type` so we save one member accessor
- If this is an event activity, `hasAttachment` should be `undefined` instead of `false`

```suggestion
        const type = activity?.type;
        const telemetryPayload: TelemetrySendStatusChangePayload = {
          clientActivityID: activity?.channelData?.clientActivityID,
          hasAttachment: type === 'message' && activity?.attachments?.length > 0 ? 'true' : 'false',
          key,
          prevStatus,
          status,
          type
        };
```

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        trackEvent && trackEvent('send-status:change', telemetryPayload);
```
Comment (compulim) [inline packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:28]: Please correct me if I am wrong. I think `trackEvent` is never falsy.

Also consider eliminate one-use constants.

```suggestion
        trackEvent('send-status:change', telemetryPayload);
```

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      }
+    }
+  }
+  return null;
```
Comment (compulim) [inline packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:32]: New line to separate block of code.

```suggestion

  return null;
```

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import { useGetActivityByKey, useSendStatusByActivityKey, useTrackEvent } from '../../hooks';
+
+import usePrevious from '../../hooks/internal/usePrevious';
+
```
Comment (compulim) [inline packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:4]: Consider typing out the payload and exporting it. We can do it in another pull request.

Notes:
- We could/should grab the `SendStatus` somewhere

```ts
type TelemetrySendStatusChangePayload = {
  clientActivityID?: string;
  hasAttachment?: 'true' | 'false';
  key: string;
  prevStatus?: 'sending' | 'send failed' | 'sent';
  status: 'sending' | 'send failed' | 'sent';
  type?: string;
};
```

File: docs/TELEMETRY.md [hunk 7]
```diff
@@ -53,9 +53,10 @@ When the following hooks are called, one or more event measurements will be emit
 
 ### Other events
 
-| Name   | Description                                |
-| ------ | ------------------------------------------ |
-| `init` | Emit when telemetry system has initialized |
+| Name                 | Description                                                                                                                                                                                                            |
+| -------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `init`               | Emit when telemetry system has initialized                                                                                                                                                                             |
+| `send-status:change` | Emit when activity status changes from `sending` to `sent`, `sending` to `send failed` and `send failed` to `sent`. Including `currentStatus`, `previousStatus`, `clientActivityID`, `key`, `hasAttachment` and `type` |
```
Comment (compulim) [inline docs/TELEMETRY.md:59]: Maybe a separate section like `HOOKS.md` that also document the type.

File: CHANGELOG.md [hunk 8]
```diff
@@ -24,6 +24,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Added
 
+-  Added function to emit status change telemetry event for activities. PR [#4631](https://github.com/microsoft/BotFramework-WebChat/pull/4631) [@Erli-ms](https://github.com/Erli-ms)
+
+### Added
+
```
Comment (compulim) [inline CHANGELOG.md:30]: I will fix the other changelog entry later.

```suggestion
-  Added function to emit status change telemetry event for activities, by [@Erli-ms](https://github.com/Erli-ms), in PR [#4631](https://github.com/microsoft/BotFramework-WebChat/pull/4631)

```

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                  telemetryEventTracking.push(event);
+                }
+              },
+            },
+            document.getElementById('webchat')
+          );
+
+          await pageConditions.uiConnected();
+
+          await directLine.emulateIncomingActivity('Aloha!');
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:37]: Let's write some comments in BDD style. I was using SETUP/WHEN/THEN. It should be GIVEN/WHEN/THEN. I will slowly fix mine.

```suggestion
          // GIVEN: An incoming activity.
          await directLine.emulateIncomingActivity('Aloha!');
```

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+          await pageConditions.became(
+            'send message successfully',
+            () => pageElements.activityStatuses()[0]?.innerText.indexOf('Just now') !== -1,
+            1000
+          );
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:43]: I think using the `numActivitiesShown()` is enough.

```suggestion
          // WHEN: The incoming message arrived.
          await pageConditions.numActivitiesShown(1);
```

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+          expect(telemetryEventTracking.length).toEqual(0);
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:45]:
```suggestion
          // THEN: No telemetry event should be emitted.
          //            This is because `send-status:change` only emit for outgoing activities, not incoming activities.
          expect(telemetryEventTracking.length).toEqual(0);
```

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        },
+        { ignoreErrors: true }
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:47]:
```suggestion
        }
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+              name: 'test.txt'
+            }
+          ]
+        }
+        const store = testHelpers.createStore();
+
+        const directLine = testHelpers.createDirectLineEmulator(store);
+
+        const telemetryEventTracking = [];
+        WebChat.renderWebChat(
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:39]: Nit.

```suggestion
        const directLine = testHelpers.createDirectLineEmulator(store);
        const telemetryEventTracking = [];

        WebChat.renderWebChat(
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+          {
+            directLine,
+            store,
+            onTelemetry: (event) => {
+              telemetryEventTracking.push(event);
+            },
+            styleOptions: {
+              sendTimeout: 5000
+            }
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:48]: Nit: sort.

```suggestion
            directLine,
            onTelemetry: (event) => {
              telemetryEventTracking.push(event);
            },
            store,
            styleOptions: {
              sendTimeout: 5000
            }
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+          },
+          document.getElementById('webchat')
+        );
+
+        await pageConditions.uiConnected();
+
+        const sendMessage1stAttempt = await directLine.emulateOutgoingActivity(testAttachmentActivity);
+        await sendMessage1stAttempt.echoBack();
+        sendMessage1stAttempt.resolvePostActivity();
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:57]: Nit. Notes:
- New line after `const` block
- `resolveAll()` is a handy version of `echoBack()` + `resolvePostActivity()`

```suggestion
        // GIVEN: An outgoing activity has sent successfully.
        const sendMessageFirstAttempt = await directLine.emulateOutgoingActivity(testAttachmentActivity);

        await sendMessageFirstAttempt.resolveAll();
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+        await pageConditions.became(
+          'successfully sent message',
+          () => {
+            return pageElements.activityStatuses()[0]?.innerText.indexOf('Just now') !== -1
+          },
+          1000
+        );
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:65]: Nit.

```suggestion
        await pageConditions.became(
          'message successfully sent',
          () => ~pageElements.activityStatuses()[0]?.innerText.indexOf('Just now'),
          1000
        );
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+        let lastRecord = telemetryEventTracking.pop();
+        expect(lastRecord.data).toEqual({
+          currentStatus: "sent",
+          previousStatus: "sending",
+          clientActivityID: expect.any(String),
+          type: "message",
+          key: expect.any(String),
+          hasAttachment: "true"
+        });
+        expect(lastRecord.name).toEqual(expectedEventName);
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:77]: Try this for better test reliability. In the future, if we added a new data point, it may break because `pop()` may not get the one we want.

Nit: sort.

[`expect.arrayContaining()` doc](https://jestjs.io/docs/expect#expectarraycontainingarray).

```suggestion
        // THEN: A "send-status:change" telemetry event should have emitted with corresponding properties.
        expect(telemetryEventTracking).toEqual(expect.arrayContaining([{
          data: {
            clientActivityID: expect.any(String),
            hasAttachment: "true"
            key: expect.any(String),
            prevStatus: "sending",
            status: "sent",
            type: "message",
          },
          name: 'send-status:change',
        }]));
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+      },
+      { ignoreErrors: true });
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:79]:
```suggestion
      });
```

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        await pageConditions.became(
+          'successfully sent message',
+          () => {
+            return pageElements.activityStatuses()[0]?.innerText.indexOf('Just now') !== -1
+          },
+          1000
+        );
+
+        const expectedEventName = 'send-status:change';
+        let lastRecord = telemetryEventTracking.pop();
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:68]: Also add an expectation for `undefined` -> `sending`.

The current expectation shows `sending` -> `sent`.

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        await pageConditions.became(
+          'failed to send message',
+          () => {
+            return pageElements.activityStatuses()[0]?.innerText === 'Send failed. Retry.' &&
+            pageElements.activityStatuses()[1]?.innerText === 'Send failed. Retry.'
+          },
+          1000
+        );
+
+        const expectedEventName = 'send-status:change';
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:49]: Add tests for `undefined` -> `sending`.

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        let lastRecord = telemetryEventTracking.pop();
+        expect(lastRecord.data).toEqual({
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:51]: Try use `expect.arrayContaining`.

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+            styleOptions: {
+              sendTimeout: 5000
+            }
+          },
+          document.getElementById('webchat')
+        );
+
+        await pageConditions.uiConnected();
+
+        const sendMessage1stAttempt = await directLine.emulateOutgoingActivity('Egress message 1, sending to send failed');
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:34]: Can you add a reason why we want 2 attempts? What cases will it additionally test when compare to only 1 attempt?

Maybe we want to check the order of the telemetry events?

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+          currentStatus: 'send failed',
+          previousStatus: 'sending',
+          clientActivityID: expect.any(String),
+          type: 'message',
+          key: expect.any(String),
+          hasAttachment: 'false'
+        });
+        expect(lastRecord.name).toEqual(expectedEventName);
+      },
+      { ignoreErrors: true });
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:72]:
```suggestion
      });
```

File: __tests__/html/activityStatusTelemetry.sendingToSendFailedToSent.html [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+        return pageConditions.became(
+          `Activity "channelData.state" should be "${expectedState}"`,
+          () => pageObjects.getActivities()[0].channelData?.state === expectedState,
+          1000
+        );
+      }
+
+      run(async function () {
+        const telemetryEventTracking = [];
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendingToSendFailedToSent.html:53]: I believe we want to use the following scenario to show `sending` -> `send failed` -> `sent`.

1. Send an activity
   - Expect `undefined` -> `sending`
2. Don't call `echoBack()` and `resolvePostActivity()`
3. Wait for 20s
   - Expect `sending` -> `send failed`
4. Call `resolveAll()` (which is `echoBack()` + `resolvePostActivity()`)
   - Expect `send failed` -> `sent`
   
You can use `lolex.install()` to create a fake timer, which you can advance its time manually.

Then, you can use `createDirectLineEmulator` without much code above.

I guess this should work. Let me know if it don't.

File: __tests__/html/activityStatusTelemetry.sendingToSent.html [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+            styleOptions: {
+              sendTimeout: 5000
+            }
+          },
+          document.getElementById('webchat')
+        );
+
+        await pageConditions.uiConnected();
+
+        const sendMessage1stAttempt = await directLine.emulateOutgoingActivity('sending to sent');
```
Comment (compulim) [inline __tests__/html/activityStatusTelemetry.sendingToSent.html:34]: See the comment for the `hasAttachment` test.

**Review by compulim — COMMENTED** [review]
Notes:
- Think if we need these in another PR:
   - In addition to `TelemetryPayload.type: 'event' | 'message' | ...`, should we also have something to tell if this is a `ImBack`/`PostBack`/`MessageBack`?
   - We should export the type, via `botframework-webchat-api/index.ts` so others know and can expect what we are sending out
      - We have 2-3 events that is already in Web Chat today, we will also need to type them out too... so in another PR or I can do it

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html (hunk 22 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:34]: The purpose is to make sure every message status update only generates one telemetry

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 26]
```diff
@@ -0,0 +1,35 @@
+import { useGetActivityByKey, useSendStatusByActivityKey, useTrackEvent } from '../../hooks';
+
+import usePrevious from '../../hooks/internal/usePrevious';
+
+const ActivitySendStatusTelemetryComposer = () => {
+  const [activityToSendStatusMap] = useSendStatusByActivityKey();
+  const previousActivityToSendStatusMap = usePrevious(activityToSendStatusMap);
```
[the remark this answers is not in the document]
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:7]: updatd

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx (hunk 1 above)
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:14]: updated

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx (hunk 2 above)
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:18]: In order to track the send status change from "undefined to sending", we need to process the if prevStatus is "undefined". changing the comment:
// `prevStatus` is undefined or a valid status, if it is undefined, it is newly added

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx (hunk 3 above)
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:27]: updated

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx (hunk 5 above)
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:32]: updated

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx (hunk 4 above)
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:28]: updated, we need to use the const of trackEvent as useTrackEvent cannot directly used in a loop. 
image

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+  const getActivityByKey = useGetActivityByKey();
+  const trackEvent = useTrackEvent();
+
+  if (activityToSendStatusMap && previousActivityToSendStatusMap) {
+    const allActivityKeys = activityToSendStatusMap.keys();
+
+    for (const key of allActivityKeys) {
+      const currentStatus = activityToSendStatusMap.get(key);
+      const previousStatus = previousActivityToSendStatusMap.get(key);
```
[the remark this answers is not in the document]
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:16]: updated

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+
+import usePrevious from '../../hooks/internal/usePrevious';
+
+const ActivitySendStatusTelemetryComposer = () => {
+  const [activityToSendStatusMap] = useSendStatusByActivityKey();
+  const previousActivityToSendStatusMap = usePrevious(activityToSendStatusMap);
+  const getActivityByKey = useGetActivityByKey();
+  const trackEvent = useTrackEvent();
+
+  if (activityToSendStatusMap && previousActivityToSendStatusMap) {
```
[the remark this answers is not in the document]
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:11]: updated

File: packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx (hunk 6 above)
Comment (Erli-ms) [reply packages/api/src/providers/ActivitySendStatusTelemetry/ActivitySendStatusTelemetryComposer.tsx:4]: add the type locally for now.

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html (hunk 9 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:37]: updated

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html (hunk 10 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:43]: updated

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html (hunk 12 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:47]: updated

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 13 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:39]: updated

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 14 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:48]: updated

File: __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html (hunk 11 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.ignoreStatusChange.incomingActivity.html:45]: updated

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 15 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:57]: updated

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 16 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:65]: removed and replaced with 
await pageConditions.numActivitiesShown(1);

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 19 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:68]: added

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 17 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html:77]: updated

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html (hunk 20 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:49]: added

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html (hunk 21 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:51]: updated

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html (hunk 23 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:72]: we need to ignore the errors as we artificially injecting errors,

File: __tests__/html/activityStatusTelemetry.sendingToSendFailedToSent.html (hunk 24 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendingToSendFailedToSent.html:53]: updated

File: __tests__/html/activityStatusTelemetry.sendingToSent.html (hunk 25 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendingToSent.html:34]: updated

File: docs/TELEMETRY.md (hunk 7 above)
Comment (Erli-ms) [reply docs/TELEMETRY.md]: added a section regarding event type below telemetry interfaces

File: __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html (hunk 18 above)
Comment (Erli-ms) [reply __tests__/html/activityStatusTelemetry.sendFiles.attachmentUrl.sendingToSent.html]: updated

File: __tests__/html/activityStatusTelemetry.sendingToSendFailed.html (hunk 23 above)
Comment (compulim) [reply __tests__/html/activityStatusTelemetry.sendingToSendFailed.html:84]: You caught me. 🤣

**Comment by compulim** [comment]
I ran tests locally/manually. Everything looks good.
